sklearn/linear_model/base.py

```diff
@@ -122,26 +120,94 @@ def center_data(X, y, fit_intercept, normalize=False, copy=True,
         if isinstance(sample_weight, numbers.Number):
             sample_weight = None
         if sp.issparse(X):
-            X_mean = np.zeros(X.shape[1])
+            X_offset = np.zeros(X.shape[1])
             X_std = np.ones(X.shape[1])
         else:
-            X_mean = np.average(X, axis=0, weights=sample_weight)
-            X -= X_mean
+            X_offset = np.average(X, axis=0, weights=sample_weight)
+            X -= X_offset
+            # XXX: currently scaled to variance=n_samples
             if normalize:
-                # XXX: currently scaled to variance=n_samples
                 X_std = np.sqrt(np.sum(X ** 2, axis=0))
                 X_std[X_std == 0] = 1
                 X /= X_std
             else:
                 X_std = np.ones(X.shape[1])
-        y_mean = np.average(y, axis=0, weights=sample_weight)
-        y = y - y_mean
+        y_offset = np.average(y, axis=0, weights=sample_weight)
+        y = y - y_offset
     else:
-        X_mean = np.zeros(X.shape[1])
+        X_offset = np.zeros(X.shape[1])
         X_std = np.ones(X.shape[1])
-        y_mean = 0. if y.ndim == 1 else np.zeros(y.shape[1], dtype=X.dtype)
-    return X, y, X_mean, y_mean, X_std
+        y_offset = 0. if y.ndim == 1 else np.zeros(y.shape[1], dtype=X.dtype)
+    return X, y, X_offset, y_offset, X_std
+
+
+def _preprocess_data(X, y, fit_intercept, normalize=False, copy=True,
+                     sample_weight=None, return_mean=False):
+    """
+    Centers data to have mean zero along axis 0. If fit_intercept=False or if
+    the X is a sparse matrix, no centering is done, but normalization can still
+    be applied. The function returns the statistics necessary to reconstruct
+    the input data, which are X_offset, y_offset, X_scale, such that the output
+
+        X = (X - X_offset) / X_scale
+
+    If sample_weight is not None, then the weighted mean of X and y
+    is zero, and not the mean itself. If return_mean=True, the mean, eventually
+    weighted, is returned, independently of whether X was centered (option used
+    for optimization with sparse data in coordinate_descend).
+
+    This is here because nearly all linear models will want their data to be
+    centered.
+    """
+
+    if isinstance(sample_weight, numbers.Number):
+        sample_weight = None
+
+    X = check_array(X, copy=copy, accept_sparse=['csr', 'csc'],
+                    dtype=FLOAT_DTYPES)
+
+    if fit_intercept:
+        if sp.issparse(X):
+            X_offset, X_var = mean_variance_axis(X, axis=0)
+            if not return_mean:
+                X_offset = np.zeros(X.shape[1])
+
+            if normalize:
+
+                # TODO: f_normalize could be used here as well but the function
+                # inplace_csr_row_normalize_l2 must be changed such that it
+                # can return also the norms computed internally
+
+                # transform variance to norm in-place
```

Comment:
We can use `csr_row_norms` which already does this, but we can leave that for later.